src/downloader/sentry_client_reactor.rs

```diff
@@ -146,6 +145,16 @@ impl SentryClientReactor {
         }
     }
 
+    pub fn send_message_queue_space_wait(
```

Comment:
This function does not send a message as far as I understand, and therefore the name could be misleading. It only gets a permit to send one, but then the permit is dropped.

Reply:
This method could actually return `OwnedPermit` to be used by invoker.

Reply:
@blasrodri what name do you suggest? is "wait_for_space_in_send_message_queue" better?

The OwnedPermit is not dropped. It is wrapped in the return value to be owned by the caller. This is the difference between `reserve` and `reserve_owned`.

Yes, I could return the OwnedPermit inside Result, but it's not needed there. If some other stage takes up the empty space by sending some messages after `FetchRequestStage.execute()`, we just gonna await again on the next loop iteration. In practice it is not happening, because for now this is the only stage that sends.

Reply:
I'm not sure I understand this part:
>  It is wrapped in the return value to be owned by the caller

When I look at the [example](https://docs.rs/tokio/1.11.0/tokio/sync/mpsc/struct.Sender.html#examples-6), the idea is to use the permit; it signals that there is a slot ready, and hence you can send something. Otherwise, you get a slot, and then that slot is freed right away when the destructor is called.

So either: 
- `send_message_queue_space_wait` sends a message after getting the permit.
- change its name to something like `probe_message_queue_for_slot`?

Reply:
With this method you don't get a slot right away, but you can wait for it by calling `.await` on the returned future. If you want to "probe" without waiting, there's another method: try_reserve/try_reserve_owned.

> you get a slot, and then that slot is freed right away when the destructor is called.
Yes, this is true, but potentially only after awaiting. The OwnedPermit is not dropped until await returns. When await returns, it is dropped, and in theory might be taken by another thread, so on the next `.execute()` iteration it will just wait again. But in practice this is the only stage that cares about sending now, so the slot remain free on the next `.execute()` iteration.

I'll rename the method to be easier to understand :+1:

Reply:
> Yes, this is true, but potentially only after awaiting. The OwnedPermit is not dropped until await returns. When await returns, it is dropped, and in theory the slot might be taken by another thread, so on the next .execute() iteration we'll just wait here again. But in practice this is the only stage that cares about sending now, so the slot remains free on the next .execute() iteration, and we can send at least 1 new request.

Looks like a footgun waiting to happen. 😕 I'd rather encode this invariant to be check by compiler than have to remember it.